components/mpas-ocean/src/shared/mpas_ocn_vmix.F

```diff
@@ -1583,9 +1583,26 @@ subroutine ocn_vmix_implicit(dt, meshPool, statePool, forcingPool, scratchPool,
                !$acc enter data copyin(tracerGroupSurfaceFlux)
 #endif
 
-               call ocn_tracer_vmix_tend_implicit(dt, vertDiffTopOfCell, layerThickness, tracersGroup, &
+               if (trim(groupItr % memberName) /= 'activeTracers' .and. &
+                   config_cvmix_background_diffusion_passive_enable) then
+                      vertDiffPassiveTopOfCell = vertDiffTopOfCell - &
+                         config_cvmix_background_diffusion + config_cvmix_background_diffusion_passive
```

Comment:
@maltrud this probably doesn't impact the simulation but this does create a nonzero diffusivity at the surface and bottom of the ocean, which I don't think we want in the code.  The change should only be from k=2:maxLevelCell

Reply:
@vanroekel and @mark-petersen thanks for finding this.  I'll clean it up--array syntax has become second nature to me so going back to loops often doesn't occur to me.

Reply:
does this look correct to you?  it compiled ok and I'll make sure it runs.

#ifdef MPAS_OPENACC
                  !$acc parallel loop gang vector collapse(3) present(vertDiffPassiveTopOfCell, vertDiffTopOfCell)
#else
                  !$omp parallel
                  !$omp do schedule(runtime) private(k, iTracer)
#endif
                  do iCell = 1, nCellsOwned
                  do k = 2, maxLevelCell(iCell)
                      vertDiffPassiveTopOfCell(k,iCell) = vertDiffTopOfCell(k,iCell) - &
                         config_cvmix_background_diffusion + config_cvmix_background_diffusion_passive
                  end do
                  end do
#ifndef MPAS_OPENACC
                  !$omp end do
                  !$omp end parallel
#endif

Reply:
should the $acc directive also have maxLevelCell?

also a general question:  in the loop @mark-petersen references as an example (thanks for that!) the k loop is 1:nVertLevels, not 1:maxLevelCell(iCell).  I realize this is a bit different (we don't want a k=1 value for diffusivity) but why not loop over 1:maxLevelCell?

Reply:
@maltrud you won't be able to collapse these loops with maxLevelCell as a loop limit. For maximum parallelism, you can either use fixed loop (2:nVertLevels) and add a conditional inside (if k < maxLevelCell) or use a mask. Otherwise, you can have just one level of parallelism over cells and leave the loop limit variable. And yes, you'll need maxLevelCell in the present() list. You also need to change collapse to either (2) in the first case or remove it if you only parallelize the cell loop. Finally, you should verify that the arrays in the present() list have all been copied to the device at some point (all mesh related variables like maxLevel are always on device and vertDiffTopOfCell I think has been copied earlier, but don't know if you added VertDiffPassive...)

Reply:
thanks for your help, @philipwjones.  I think I'll opt for limits of 2:nVertLevels and adding a conditional, though I assume that would affect vectorization.  would using 'merge' help for vectorization?  eg:

vertDiffPassiveTopOfCell(k,iCell) = merge( vertDiffTopOfCell(k,iCell) -....,0.0_RKIND,k<maxLevelCell(k,iCell))

how can I tell if something is present?  I see $acc update host for vertDiffTopOfCell in ocn_vmix_coefs, then $acc update device after the coefficients are calculated.  do I need to do something similar for vertDiffPassiveTopOfCell?  I did put in $acc update host for vertDiffPassiveTopOfCell but I suspect that isn't correct?

Reply:
@maltrud it may affect CPU/vector performance but I suspect this loop has a fairly small cost anyway. The merge probably won't be any better - this is a simple enough loop the compiler should figure it out anyway and sometimes the intrinsic is actually worse.  I didn't look closely at the data transfers when I wrote that last, but what you have looks correct - this is computed on the device so you just need to update the host which you've done. And you added it to the create/destroy calls in diagnostic_vars so should be fine.